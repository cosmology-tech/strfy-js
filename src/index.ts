export * from './js';
export * from './json';