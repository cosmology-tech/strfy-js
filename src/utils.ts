import { minimatch } from 'minimatch';

export function camelCaseTransform(key: string): string {
  return key.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
}

export function isSimpleKey(key: string): boolean {
  return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key);
}

export function dirname(path: string): string {
  return path.replace(/\/[^\/]*$/, ''); // Removes last segment after the last '/'
}

export function escapeStringForSingleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only single quotes
  str = str.replace(/'/g, "\\'");
  return str;
}

export function escapeStringForDoubleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only double quotes
  str = str.replace(/"/g, '\\"');
  return str;
}


export function escapeStringForBacktickQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape backticks
  str = str.replace(/`/g, '\\`');
  return str;
}


const globPattern = /\*+([^+@!?\*\[\(]*)/;

interface ShouldIncludeOptions {
  include: string[];
  exclude: string[];
}

export function expandPaths(paths: string[]): string[] {
  const pathSet = new Set<string>();

  paths.forEach(path => {
    let currentPath = path;
    // Ensure the path is added to the set
    pathSet.add(currentPath);

    // Generate and add all parent directories to the set
    while (currentPath !== '') {
      const lastSlashIndex = currentPath.lastIndexOf('/');
      if (lastSlashIndex === -1) break; // No more slashes, stop processing

      currentPath = currentPath.substring(0, lastSlashIndex) || '';
      if (currentPath !== '')
        pathSet.add(currentPath);
    }
  });

  return sortPaths(Array.from(pathSet));
}

function sortPaths(paths) {
  // First sort alphabetically
  paths.sort((a, b) => a.localeCompare(b));

  // Then sort by depth (number of '/' in the path)
  paths.sort((a, b) => {
    const depthA = (a.match(/\//g) || []).length;
    const depthB = (b.match(/\//g) || []).length;
    return depthB - depthA;  // Sort from deepest to shallowest
  });

  return paths;
}


function hasAncestor(currentPath, includePaths) {
  return !!findNearestAncestor(currentPath, includePaths);
}

function findNearestAncestor(currentPath, includePaths) {
  let pathToCheck = currentPath;
  while (pathToCheck !== '/' && pathToCheck !== '') {
    if (includePaths.some(includePath => minimatch(pathToCheck + '/', includePath))) {
      if (pathToCheck !== currentPath)
        return pathToCheck; // Return the matching parent directory
    }
    pathToCheck = dirname(pathToCheck); // Move to the next parent directory
  }
  return null; // Return null if no matching parent is found
}

export const shouldInclude = (type: string, options: ShouldIncludeOptions): boolean => {
  // Determine if 'include' and 'exclude' are effectively set
  const includesEffectivelySet = options.include && options.include.length > 0;
  const excludesEffectivelySet = options.exclude && options.exclude.length > 0;

  // Function to check if any patterns in the array match the type
  const matchesPattern = (patterns: string[], type: string): boolean =>
    patterns.some(pattern => globPattern.test(pattern) ? minimatch(type, pattern) : type === pattern);

  const isIncluded = includesEffectivelySet ? matchesPattern(expandPaths(options.include), type) : true;
  const isIncludedByAncestor = !!findNearestAncestor(type, options.include);
  const isExcluded = excludesEffectivelySet ? matchesPattern(options.exclude, type) : false;

  const toInclude = isIncluded;

  // Apply the logic based on whether includes or excludes are effectively set
  if (includesEffectivelySet && excludesEffectivelySet) {
    if (isIncludedByAncestor && !isExcluded) {
      return true;
    }
    return toInclude && !isExcluded;
  } else if (includesEffectivelySet) {
    if (isIncludedByAncestor) {
      return true;
    }
    return toInclude;
  } else if (excludesEffectivelySet) {
    return !isExcluded;
  }

  // Default behavior if neither is effectively set
  return true;
}