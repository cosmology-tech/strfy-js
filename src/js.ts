import { camelCaseTransform, escapeStringForBacktickQuotes, escapeStringForDoubleQuotes, escapeStringForSingleQuotes, isSimpleKey } from "./utils";

export interface JSStringifyOptions {
  space?: number;
  replacer?: (key: string, value: any) => any | null;
  propertyReplacer?: (currentKey: string, currentPath: string, propertyRenameMap: { [pattern: string]: string }) => string
  quotes?: 'single' | 'double' | 'backtick';
  inlineArrayLimit?: number;
  camelCase?: boolean;
  camelCaseFn?: (str: string) => string;
  propertyRenameMap?: { [key: string]: string };
  json?: boolean;
}


export function chooseQuotes(str: string, preferred: 'single' | 'double' | 'backtick'): string {
  switch (preferred) {
    case 'single':
      return `'${escapeStringForSingleQuotes(str)}'`;
    case 'double':
      return `"${escapeStringForDoubleQuotes(str)}"`;
    case 'backtick':
      return `\`${escapeStringForBacktickQuotes(str)}\``;
    default:
      throw new Error("Invalid quote type specified.");
  }
}

function matchesPattern(path, pattern) {
  const pathSegments = path.split('/');
  const patternSegments = pattern.split('/');
  if (pathSegments.length !== patternSegments.length) {
    return false;
  }
  for (let i = 0; i < pathSegments.length; i++) {
    if (patternSegments[i] !== '*' && patternSegments[i] !== pathSegments[i]) {
      return false;
    }
  }
  return true;
}

function defaultPropertyReplacer(
  currentKey: string,
  currentPath: string,
  propertyRenameMap: { [pattern: string]: string }
): string {
  let finalKey = currentKey;
  Object.keys(propertyRenameMap).forEach(pattern => {
    if (matchesPattern(currentPath, pattern)) {
      finalKey = propertyRenameMap[pattern];
    }
  });
  return finalKey;
}


export function jsStringify(obj: any, options?: JSStringifyOptions): string {
  const {
    space = 0,
    replacer = null,
    propertyReplacer = null,
    quotes = 'single',
    inlineArrayLimit = undefined,
    camelCase = false,
    camelCaseFn = camelCaseTransform,
    propertyRenameMap = {},
    json = false  // JSON compliance mode
  } = options || {};

  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);

  let indentLevel: number = 0;

  const serialize = (obj: any, currentPath = "", isArray: boolean = false): string => {
    if (replacer instanceof Function) {
      obj = replacer.call(null, '', obj); // Call replacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map((item, index) => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(item, `${currentPath}/${index}`, true);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;
      const props = Object.keys(obj).map(key => {
        const fullPath = `${currentPath}/${key}`;
        const value = replacer instanceof Function ? replacer.call(obj, key, obj[key]) : obj[key];


        let replacedKey: string;
        switch (true) {
          case propertyReplacer instanceof Function:
            replacedKey = propertyReplacer.call(obj, key, fullPath, propertyRenameMap);
            break;
          default:
            replacedKey = defaultPropertyReplacer(key, fullPath, propertyRenameMap);
        }

        const finalKey = camelCase ? camelCaseFn(replacedKey) : replacedKey;
        const keyPart = json ? `"${finalKey}"` : isSimpleKey(finalKey) ? finalKey : `"${finalKey}"`;
        const valuePart = serialize(value, fullPath);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return json ? `"${escapeStringForDoubleQuotes(obj)}"` : chooseQuotes(obj, quotes);
    } else {
      return String(obj);
    }
  };

  return serialize(obj);
}
