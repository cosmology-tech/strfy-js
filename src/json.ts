import { jsStringify, JSStringifyOptions } from "./js";

export interface JSONStringifyOptions extends Omit<JSStringifyOptions, 'json'> { }

export function jsonStringify(obj: any, options?: JSONStringifyOptions): string {
  // Set json option to true and spread the rest of the options
  return jsStringify(obj, { ...options, json: true });
}
