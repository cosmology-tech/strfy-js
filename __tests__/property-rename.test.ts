import { minimatch } from "minimatch";
import { jsStringify } from "../src";

const assetlist = {
  "$schema": "../assetlist.schema.json",
  "chain_name": "comdex",
  "assets": [
    {
      "description": "Native Token of Comdex Protocol",
      "denom_units": [
        {
          "denom": "ucmdx",
          "exponent": 0
        },
        {
          "denom": "cmdx",
          "exponent": 6
        }
      ],
      "base": "ucmdx",
      "name": "Comdex",
      "display": "cmdx",
      "symbol": "CMDX",
      "logo_URIs": {
        "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png",
        "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg"
      },
      "coingecko_id": "comdex",
      "images": [
        {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg"
        }
      ]
    },
    {
      "type_asset": "sdk.Coin",
      "description": "Governance Token of Harbor protocol on Comdex network",
      "denom_units": [
        {
          "denom": "uharbor",
          "exponent": 0
        },
        {
          "denom": "harbor",
          "exponent": 6
        }
      ],
      "base": "uharbor",
      "name": "Harbor",
      "display": "harbor",
      "symbol": "HARBOR",
      "logo_URIs": {
        "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png",
        "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg"
      },
      "coingecko_id": "harbor-2",
      "images": [
        {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg"
        }
      ]
    },
    {
      "description": "Stable Token of Harbor protocol on Comdex network",
      "denom_units": [
        {
          "denom": "ucmst",
          "exponent": 0
        },
        {
          "denom": "cmst",
          "exponent": 6
        }
      ],
      "base": "ucmst",
      "name": "CMST",
      "display": "cmst",
      "symbol": "CMST",
      "logo_URIs": {
        "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png",
        "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg"
      },
      "coingecko_id": "composite",
      "images": [
        {
          "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png",
          "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg"
        }
      ]
    }
  ]
};


it('assetlist', () => {

  const options = {
    camelCase: true,
    space: 2,
    propertyRenameMap: {
      "/assets/*/denom_units": "denominations",
      "/assets/[0-1]/logo_URIs": "logos"
    }
  };

  const jsonString = jsStringify(assetlist, options);

  expect(jsonString).toMatchSnapshot();
});

function propertyReplacer(
  currentKey: string,
  currentPath: string,
  propertyRenameMap: { [pattern: string]: string }
): string {
  let finalKey = currentKey;
  Object.keys(propertyRenameMap).forEach(pattern => {
    if (minimatch(currentPath, pattern)) {
      finalKey = propertyRenameMap[pattern];
    }
  });
  return finalKey;
}

it('replacer', () => {
  const options = {
    camelCase: true,
    space: 2,
    propertyRenameMap: {
      "/assets/*/denom_units": "denominations",
      "/assets/[0-1]/logo_URIs": "logos"
    },
    propertyReplacer
  };
  const jsonString = jsStringify(assetlist, options);
  expect(jsonString).toMatchSnapshot();
});