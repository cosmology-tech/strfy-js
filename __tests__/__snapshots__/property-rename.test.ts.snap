// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assetlist 1`] = `
"{
  $schema: '../assetlist.schema.json',
  chainName: 'comdex',
  assets: [
    {
      description: 'Native Token of Comdex Protocol',
      denominations: [
        {
          denom: 'ucmdx',
          exponent: 0
        },
        {
          denom: 'cmdx',
          exponent: 6
        }
      ],
      base: 'ucmdx',
      name: 'Comdex',
      display: 'cmdx',
      symbol: 'CMDX',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg'
      },
      coingeckoId: 'comdex',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg'
        }
      ]
    },
    {
      typeAsset: 'sdk.Coin',
      description: 'Governance Token of Harbor protocol on Comdex network',
      denominations: [
        {
          denom: 'uharbor',
          exponent: 0
        },
        {
          denom: 'harbor',
          exponent: 6
        }
      ],
      base: 'uharbor',
      name: 'Harbor',
      display: 'harbor',
      symbol: 'HARBOR',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg'
      },
      coingeckoId: 'harbor-2',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg'
        }
      ]
    },
    {
      description: 'Stable Token of Harbor protocol on Comdex network',
      denominations: [
        {
          denom: 'ucmst',
          exponent: 0
        },
        {
          denom: 'cmst',
          exponent: 6
        }
      ],
      base: 'ucmst',
      name: 'CMST',
      display: 'cmst',
      symbol: 'CMST',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
      },
      coingeckoId: 'composite',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
        }
      ]
    }
  ]
}"
`;

exports[`replacer 1`] = `
"{
  $schema: '../assetlist.schema.json',
  chainName: 'comdex',
  assets: [
    {
      description: 'Native Token of Comdex Protocol',
      denominations: [
        {
          denom: 'ucmdx',
          exponent: 0
        },
        {
          denom: 'cmdx',
          exponent: 6
        }
      ],
      base: 'ucmdx',
      name: 'Comdex',
      display: 'cmdx',
      symbol: 'CMDX',
      logos: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg'
      },
      coingeckoId: 'comdex',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg'
        }
      ]
    },
    {
      typeAsset: 'sdk.Coin',
      description: 'Governance Token of Harbor protocol on Comdex network',
      denominations: [
        {
          denom: 'uharbor',
          exponent: 0
        },
        {
          denom: 'harbor',
          exponent: 6
        }
      ],
      base: 'uharbor',
      name: 'Harbor',
      display: 'harbor',
      symbol: 'HARBOR',
      logos: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg'
      },
      coingeckoId: 'harbor-2',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg'
        }
      ]
    },
    {
      description: 'Stable Token of Harbor protocol on Comdex network',
      denominations: [
        {
          denom: 'ucmst',
          exponent: 0
        },
        {
          denom: 'cmst',
          exponent: 6
        }
      ],
      base: 'ucmst',
      name: 'CMST',
      display: 'cmst',
      symbol: 'CMST',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
      },
      coingeckoId: 'composite',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
        }
      ]
    }
  ]
}"
`;
